src/query/storage/promremote/storage.go

```diff
@@ -0,0 +1,158 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/uber-go/tally"
+
+	"github.com/m3db/m3/src/query/storage"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+)
+
+const metricsScope = "prom_remote_storage"
+
+// NewStorage returns new Prometheus remote write compatible storage
+func NewStorage(opts Options) (storage.Storage, error) {
+	client := xhttp.NewHTTPClient(opts.httpOptions)
+	scope := opts.scope.SubScope(metricsScope)
+	s := &promStorage{opts: opts, client: client, endpointMetrics: initEndpointMetrics(opts.endpoints, scope)}
+	return s, nil
+}
+
+type promStorage struct {
+	opts            Options
+	client          *http.Client
+	endpointMetrics map[string]*instrument.MethodMetrics
+}
+
+func (p *promStorage) Write(ctx context.Context, query *storage.WriteQuery) error {
+	encoded, err := encodeWriteQuery(query)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	multiErr := xerrors.NewMultiError()
+	var errLock sync.Mutex
+	for _, endpoint := range p.opts.endpoints {
+		endpoint := endpoint
+		if endpoint.resolution == query.Attributes().Resolution &&
+			endpoint.retention == query.Attributes().Retention {
+			metrics := p.endpointMetrics[endpoint.name]
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				err := p.writeSingle(ctx, metrics, endpoint.address, bytes.NewBuffer(encoded))
+				if err != nil {
+					errLock.Lock()
+					multiErr = multiErr.Add(err)
+					errLock.Unlock()
+					return
+				}
+			}()
+		}
+	}
+
+	wg.Wait()
+
+	if !multiErr.Empty() {
+		return multiErr
+	}
+	return nil
+}
+
+func (p *promStorage) Type() storage.Type {
+	return storage.TypeRemoteDC
+}
+
+func (p *promStorage) Close() error {
+	p.client.CloseIdleConnections()
+	return nil
+}
+
+func (p *promStorage) ErrorBehavior() storage.ErrorBehavior {
+	return storage.BehaviorFail
+}
+
+func (p *promStorage) Name() string {
+	return "prom-remote"
+}
+
+func (p *promStorage) writeSingle(
+	ctx context.Context,
+	metrics *instrument.MethodMetrics,
+	address string,
+	encoded io.Reader,
+) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, address, encoded)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("content-encoding", "snappy")
+	req.Header.Set(xhttp.HeaderContentType, xhttp.ContentTypeProtobuf)
+
+	start := time.Now()
+	resp, err := p.client.Do(req)
+	methodDuration := time.Since(start)
+	if err != nil {
+		metrics.ReportError(methodDuration)
+		return err
+	}
+	metrics.ReportSuccess(methodDuration)
+	defer func() { _ = resp.Body.Close() }()
```

Comment:
Nit:
```suggestion
	metrics.ReportSuccessOrError(err, time.Since(start))
	if err != nil {
		return err
	}

	defer func() { _ = resp.Body.Close() }()
```

Reply:
I actually noticed that success was being recorded in case non 200 http status was returned. Fixed that.